### PR TITLE
refactor(tm2): make rpc client not depend on goleveldb

### PR DESCRIPTION
**Pull request by thehowl**
image

image

The PR depends on #1602 but it's good for review. For ease of reviewing I put that PR's branch as the base branch; once it is merged I will rebase on master.

A bunch of changes to get there:

- Split `proxy` into `appconn` and `proxy`. `proxy` has a dependency on `kvstore`, which then depends on pkg/db and goleveldb.
- Tools for inspecting dependency chains:
  - `go list -f '{{ join .Deps "\n" }}' github.com/gnolang/gno/tm2/your/pkg`
  - [`depth`](https://github.com/KyleBanks/depth)

The original point of the PR was to not include an entire key-store database as part of `cmd/gno`, but really the better side-effect is that the dependency is no longer in the rpc package. Sadly, `gnoclient` still imports `goleveldb`, but I'll leave understanding why that is the case for another refactor-investigation :)

### CODE REVIEW

File: tm2/pkg/bft/consensus/wal_test.go [hunk 1]
```diff
@@ -84,3 +99,105 @@ func TestWALTruncate(t *testing.T) {
...
+
+// XXX: WALGenerateNBlocks and WALWithNBlocks were removed from wal_generator.go
+// as they are unused.
+// If you intend to use them, please move them to a separate package.
+
+// WALGenerateNBlocks generates a consensus WAL. It does this by spinning up a
+// stripped down version of node (proxy app, event bus, consensus state) with a
+// persistent kvstore application and special consensus wal instance
+// (heightStopWAL) and waits until numBlocks are created. If the node fails to produce given numBlocks, it returns an error.
+func WALGenerateNBlocks(t *testing.T, wr io.Writer, numBlocks int) (err error) {
```
Comment (zivkovicmilos) [inline tm2/pkg/bft/consensus/wal_test.go:111]: Thank you for moving this out 🙏

File: tm2/pkg/bft/types/keys.go [hunk 2]
```diff
@@ -1,6 +1,5 @@
 package types
 
 // UNSTABLE
-var (
-	PeerStateKey = "ConsensusReactor.peerState"
-)
```
Comment (zivkovicmilos) [inline tm2/pkg/bft/types/keys.go:6]: How this was not caught by the linters (more than 2 that check for this) is beyond me

**Review by zivkovicmilos — APPROVED** [review]
Thank you for resolving this coupling 🙏 

Looks great 💯

**Review by gfanton — APPROVED** [review]

**Review by albttx — APPROVED** [review]

**Comment by piux2** [comment]
The original TM2 allows for in-process app and RPC versus separate server configurations. With this change, can we still deploy TM2, the app, and RPC as separate services in production?

**Comment by thehowl** [comment]
@piux2 The change only affects the RPC _client_. No breaking change was introduced, and as you mentioned, we can still have separate tm2/app/rpc.

**Comment by piux2** [comment]
> @piux2 The change only affects the RPC _client_. No breaking change was introduced, and as you mentioned, we can still 
have separate tm2/app/rpc server.

Thank you for the clarification. I'm not sure I fully understand the intentions behind this PR for refactoring with all these changes

Could you help to clarify a few points?

1) Why do we want to change this interface?

```
func NewLocal(node *nm.Node) *Local {
```

https://github.com/gnolang/gno/pull/1603/files#diff-40f9edc97cafeef3b08870b23f9c1726a7ca1f624b949194b9dfac2188f58346L39

2) Why was this function removed?

```
func (ps *PeerState) ToJSON()
```

https://github.com/gnolang/gno/pull/1603/files#diff-cbf5dcdca399dbb992b6443cfec33d11f3dd043771025ce250df2398c3113a5bL933

3) Regarding the following line of code: 
```
rpccore.SetGetFastSync(n.consensusReactor.FastSync)
```
Why do we set code FastSync and remove setting the consensusReactor for the RPC?

https://github.com/gnolang/gno/pull/1603/files#diff-83913b6c6e92b6851951c0f50d813df7becd09b5cd235dabfcbc584be15941f8R700

4) The state of `consensusReactor.FastSync()` could change after the chain starts. 

`if consensusReactor.FastSync() { // could have different result `

The current implementation only use the on-start state of the consensusReactor

`if getFastSync() { // this does  not reflect the latest states of consensusReactor`

https://github.com/gnolang/gno/pull/1603/files#diff-93f50c3c46fcb420642932a285a68fa1cd3dc696578a95084115896c1d1bc93cL77

File: tm2/pkg/bft/node/node.go [hunk 3]
```diff
@@ -588,6 +589,16 @@ func (n *Node) OnStart() error {
 		time.Sleep(genTime.Sub(now))
 	}
 
+	// Set up the GLOBAL variables in rpc/core which refer to this node.
```
Comment (jaekwon) [inline tm2/pkg/bft/node/node.go:592]: ok!

File: tm2/pkg/bft/rpc/client/localclient.go [hunk 4]
```diff
@@ -3,41 +3,40 @@ package client
...
-are compiled in process.
-
-For real clients, you probably want to use client.HTTP.  For more
-powerful control during testing, you probably want the "client/mock" package.
-*/
+// Local is a Client implementation that directly executes the rpc
+// functions on a given node, without going through any network connection.
+//
+// As this connects directly to a Node instance, a Local client only works
+// after the Node has been started. Note that the way this works is (alas)
```
Comment (jaekwon) [inline tm2/pkg/bft/rpc/client/localclient.go:17]: +1

**Comment by jaekwon** [comment]
Thank you Morgan, good work!

**Comment by thehowl** [comment]
@piux2 please add inline comments next time, it makes answering your questions easier :)

> Could you help to clarify a few points?
> 
>     1. Why do we want to change this signature?
> 
> 
> ```
> func NewLocal(node *nm.Node) *Local {
> ```
> 
> https://github.com/gnolang/gno/pull/1603/files#diff-40f9edc97cafeef3b08870b23f9c1726a7ca1f624b949194b9dfac2188f58346L39

Removes the dependency from rpc/client to `github.com/gnolang/gno/tm2/pkg/bft/node`. `pkg/bft/node` depends on `goleveldb` in this way:

```
$ depth -explain github.com/gnolang/gno/tm2/pkg/db/goleveldb .
. -> github.com/gnolang/gno/tm2/pkg/bft/consensus -> github.com/gnolang/gno/tm2/pkg/bft/proxy -> github.com/gnolang/gno/tm2/pkg/bft/abci/example/kvstore -> github.com/gnolang/gno/tm2/pkg/db/goleveldb
```

Practically, we're separating concerns between what client/rpc should do. Rather than also set up the connections to make the local client work, rpc/client only provides the client itself, and the responsibility to wire everything up is with the caller.

>     2. Why was this function removed?
> 
> 
> ```
> func (ps *PeerState) ToJSON()
> ```
> 
> https://github.com/gnolang/gno/pull/1603/files#diff-cbf5dcdca399dbb992b6443cfec33d11f3dd043771025ce250df2398c3113a5bL933

[See rpc/core/consensus.go](https://github.com/gnolang/gno/pull/1603/files#diff-630ab5b18ba3a6424db79d28b5d6ffd639ef63acc3041140e05d0ec26f57e4d8). `rpc/core` is imported by the RPC client. The function `DumpConsensusState` will print the peer state for each peer. There are other places where we get PeerStateKey, so I didn't want to completely refactor it.

However, since only rpc/core with this function will need the function to print it as JSON, I added a separate type in `bft/consensus/types` with the fields that should, indeed, be exposed. This avoids the import on `bft/consensus`, which eventually imports `goleveldb` (see the above `depth` result).

>     3. Regarding the following line of code:
> 
> 
> ```
> rpccore.SetGetFastSync(n.consensusReactor.FastSync)
> ```
> 
> Why do we set code FastSync and remove setting the consensusReactor for the RPC?
> 
> https://github.com/gnolang/gno/pull/1603/files#diff-83913b6c6e92b6851951c0f50d813df7becd09b5cd235dabfcbc584be15941f8R700
> 
>     4. The state of `consensusReactor.FastSync()` could change after the chain starts.
> 
> 
> `if consensusReactor.FastSync() { // could have different result `
> 
> The current implementation only use the on-start state of the consensusReactor
> 
> `if getFastSync() { // this does not reflect the latest states of consensusReactor`
> 
> https://github.com/gnolang/gno/pull/1603/files#diff-93f50c3c46fcb420642932a285a68fa1cd3dc696578a95084115896c1d1bc93cL77
> 
>     5. Why do we make these changes?
>        https://github.com/gnolang/gno/pull/1603/files#diff-630ab5b18ba3a6424db79d28b5d6ffd639ef63acc3041140e05d0ec26f57e4d8L220

**Comment by thehowl** [comment]
@piux2 


>     3. Regarding the following line of code:
> 
> 
> ```
> rpccore.SetGetFastSync(n.consensusReactor.FastSync)
> ```
> 
> Why do we set code FastSync and remove setting the consensusReactor for the RPC?
> 
> https://github.com/gnolang/gno/pull/1603/files#diff-83913b6c6e92b6851951c0f50d813df7becd09b5cd235dabfcbc584be15941f8R700

Note that in `rpccore.SetGetFastSync(n.consensusReactor.FastSync)`, `n.consensusReactor.FastSync` is a method (a getter, specifically). The reasoning here is similar to the above: `rpc/core` should not depend on `bft/consensus`, but because it is also the hub through which things are exposed in RPC it should have access to the node's properties.

To make it have access to the properties, while not importing `bft/consensus`, we change the `consensusReactor` to be a closure: `getFastSync func() bool`. This also should help answer your next question...

>     4. The state of `consensusReactor.FastSync()` could change after the chain starts.
> 
> 
> `if consensusReactor.FastSync() { // could have different result `
> 
> The current implementation only use the on-start state of the consensusReactor
> 
> `if getFastSync() { // this does not reflect the latest states of consensusReactor`
> 
> https://github.com/gnolang/gno/pull/1603/files#diff-93f50c3c46fcb420642932a285a68fa1cd3dc696578a95084115896c1d1bc93cL77

Not really.

`getFastSync` is a _getter_. Because it is a method with a pointer receiver, it will return the value of `fastSync` for the node at the given pointer. Since [this](https://github.com/gnolang/gno/pull/1603/files#diff-93f50c3c46fcb420642932a285a68fa1cd3dc696578a95084115896c1d1bc93cL77) is the only place we used `consensusReactor`, using the getter will return the up-to-date value.

I was considering making it a property, but set for making it a getter for the precise reason it could be updated during a node's lifespan.

> 
>     5. Why do we make these changes?
>        https://github.com/gnolang/gno/pull/1603/files#diff-630ab5b18ba3a6424db79d28b5d6ffd639ef63acc3041140e05d0ec26f57e4d8L220

See answer number 2.

**Pull request merged**